frameworks.qmd

```diff
@@ -1,119 +1,1793 @@
+    > platform designed to deploy and manage machine learning pipelines
+    > in production settings. TFX encompasses components for data
+    > validation, preprocessing, model training, validation, and
+    > serving.
+
+Source:
+[[https://www.tensorflow.org/learn]{.underline}](https://www.tensorflow.org/learn)
+
+TensorFlow was developed to address the limitations of DistBelief[^2]---the
+framework in use at Google from 2011 to 2015---by providing flexibility
+along three axes: 1) defining new layers, 2) refining training
+algorithms, and 3) defining new training algorithms. To understand what
+limitations in DistBelief led to the development of TensorFlow, we will
+first give a brief overview of the Parameter Server Architecture that
+DistBelief employed[^3].
+
+The Parameter Server (PS) architecture is a popular design for
+distributing the training of machine learning models, especially deep
+neural networks, across multiple machines. The fundamental idea is to
+separate the storage and management of model parameters from the
+computation used to update these parameters:
+
+**Storage**: The storage and management of model parameters were handled
+by the stateful parameter server processes. Given the large scale of
+models and the distributed nature of the system, these parameters were
+sharded across multiple parameter servers. Each server maintained a
+portion of the model parameters, making it \"stateful\" as it had to
+maintain and manage this state across the training process.
+
+**Computation**: The worker processes, which could be run in parallel,
+were stateless and purely computational, processing data and computing
+gradients without maintaining any state or long-term memory[^4].
+
+DistBelief and its architecture defined above were crucial in enabling
+distributed deep learning at Google but also introduced limitations that
+motivated the development of TensorFlow:
+
+### Static Computation Graph
+
+In the parameter server architecture, model parameters are distributed
+across various parameter servers. Since DistBelief was primarily
+designed for the neural network paradigm, parameters corresponded to a
+fixed structure of the neural network. If the computation graph were
+dynamic, the distribution and coordination of parameters would become
+significantly more complicated. For example, a change in the graph might
+require the initialization of new parameters or the removal of existing
+ones, complicating the management and synchronization tasks of the
+parameter servers. This made it harder to implement models outside the
+neural framework or models that required dynamic computation graphs.
+
+TensorFlow was designed to be a more general computation framework[^2] where
+the computation is expressed as a data flow graph. This allows for a
+wider variety of machine learning models and algorithms outside of just
+neural networks, and provides flexibility in refining models.
+
```

Comment:
I think it could be helpful to give more context on how exactly a dynamic computation graph works! This reference might be useful: https://www.geeksforgeeks.org/dynamic-vs-static-computational-graphs-pytorch-and-tensorflow/#